cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/AbstractInMemoryProgramRunner.java

```diff
@@ -130,17 +133,49 @@ private ProgramOptions createComponentOptions(String name, int instanceId, int i
     private final ProgramOptions options;
     private final Lock lock = new ReentrantLock();
     private final ProgramRunnerFactory.Type type;
+    private final AtomicLong numWorkerThreads;
 
     public InMemoryProgramController(Table<String, Integer, ProgramController> components,
-                              RunId runId, Program program, ProgramSpecification spec,
-                              ProgramOptions options, ProgramRunnerFactory.Type type) {
+                                     RunId runId, Program program, ProgramSpecification spec,
+                                     ProgramOptions options, ProgramRunnerFactory.Type type) {
       super(program.getName(), runId);
       this.program = program;
       this.components = components;
       this.spec = spec;
       this.options = options;
       this.type = type;
+      this.numWorkerThreads = new AtomicLong(components.size());
       started();
+      monitorComponents();
+    }
+
+    // Add listener to monitor completion/killed status of individual components, so that the program can be marked
+    // as completed once all the components have completed/killed.
+    private void monitorComponents() {
+      for (ProgramController controller : components.values()) {
+        controller.addListener(new AbstractListener() {
+          @Override
+          public void completed() {
+            super.completed();
+            long value = numWorkerThreads.decrementAndGet();
+            checkThreadCompleted(value);
+          }
+
+          @Override
+          public void killed() {
+            super.killed();
+            long value = numWorkerThreads.decrementAndGet();
+            checkThreadCompleted(value);
```

Comment:
Why need to handle killed? Isn't kill is from external stop through this ProgramController?